File: scrimage-core/src/main/scala/com/sksamuel/scrimage/canvas/CaptionFilter.scala
package com.sksamuel.scrimage.canvas

import java.awt.{AlphaComposite, Graphics2D}

import com.sksamuel.scrimage.Position.BottomLeft
import com.sksamuel.scrimage.canvas.drawable.{FilledRect, Text}
import com.sksamuel.scrimage.{Color, Filter, Image, Position}

import scala.language.implicitConversions

class CaptionFilter(text: String,
                    textSize: Int = 18,
                    position: Position = BottomLeft,
                    x: Int = -1,
                    y: Int = -1,
                    font: Font = Font.as(java.awt.Font.SANS_SERIF),
                    textColor: Color = Color.White,
                    textAlpha: Double = 0.1d,
                    antiAlias: Boolean = true,
                    fullWidth: Boolean = false,
                    captionBackground: Color = Color.White,
                    captionAlpha: Double = 0.1,
                    padding: Padding = Padding(10)) extends Filter {
  require(textSize > 0, "Font size must be > 0")

  def apply(image: Image): Unit = {

    val g2 = image.awt.getGraphics.asInstanceOf[Graphics2D]
    g2.setFont(new java.awt.Font(font.name, 0, textSize))

    val bounds = g2.getFontMetrics.getStringBounds(text, g2)
    val descent = g2.getFontMetrics.getDescent
    val captionWidth = if (fullWidth) image.width else bounds.getWidth.toInt + padding.left + padding.right
    val captionHeight = bounds.getHeight.toInt + padding.top + padding.bottom

    // captionx/y are the top/left coordinates for the caption box
    val (captionX, captionY) = if (x == -1 || y == -1) {
      position.calculateXY(image.width, image.height, captionWidth, captionHeight)
    } else {
      (x, y)
    }

    import Canvas._

    val bg = FilledRect(
      captionX,
      image.height - captionHeight,
      captionWidth,
      captionHeight,
      g2 => {
        g2.setComposite(AlphaComposite.getInstance(AlphaComposite.SRC_OVER, captionAlpha.toFloat))
        g2.setFont(new java.awt.Font(font.wrapped.getName, 0, textSize))
        g2.setColor(captionBackground)
        g2.setAntiAlias(antiAlias)
      }
    )

    val string = Text(
      text,
      captionX + padding.left,
      captionY + padding.top + g2.getFontMetrics.getHeight - descent,
      g2 => {
        g2.setComposite(AlphaComposite.getInstance(AlphaComposite.SRC_OVER, textAlpha.toFloat))
        g2.setColor(textColor)
        g2.setAntiAlias(antiAlias)
      }
    )

    image.drawInPlace(bg, string)
  }
}

File: scrimage-filters/src/main/java/com/sksamuel/scrimage/filter/CaptionFilter.java
package com.sksamuel.scrimage.filter;

import com.sksamuel.scrimage.Dimension;
import com.sksamuel.scrimage.Filter;
import com.sksamuel.scrimage.Graphics2DUtils;
import com.sksamuel.scrimage.ImmutableImage;
import com.sksamuel.scrimage.Position;
import com.sksamuel.scrimage.canvas.Canvas;
import com.sksamuel.scrimage.canvas.drawables.FilledRect;
import com.sksamuel.scrimage.canvas.drawables.Text;

import java.awt.AlphaComposite;
import java.awt.Color;
import java.awt.Font;
import java.awt.Graphics2D;
import java.awt.geom.Rectangle2D;

public class CaptionFilter implements Filter {

    private final String text;
    private final Position position;
    private final int x;
    private final int y;
    private final Font font;
    private final Color textColor;
    private final double textAlpha;
    private final boolean antiAlias;
    private final boolean fullWidth;
    private final Color captionBackground;
    private final double captionAlpha;
    private final Padding padding;

    public CaptionFilter(String text, Position position, Font font, Color textColor, double textAlpha, boolean antiAlias, boolean fullWidth, Color captionBackground, double captionAlpha, Padding padding) {
        this.text = text;
        this.position = position;
        this.x = -1;
        this.y = -1;
        this.font = font;
        this.textColor = textColor;
        this.textAlpha = textAlpha;
        this.antiAlias = antiAlias;
        this.fullWidth = fullWidth;
        this.captionBackground = captionBackground;
        this.captionAlpha = captionAlpha;
        this.padding = padding;
    }

    public CaptionFilter(String text, int x, int y, Font font, Color textColor, double textAlpha, boolean antiAlias, boolean fullWidth, Color captionBackground, double captionAlpha, Padding padding) {
        this.text = text;
        this.position = null;
        this.x = x;
        this.y = y;
        this.font = font;
        this.textColor = textColor;
        this.textAlpha = textAlpha;
        this.antiAlias = antiAlias;
        this.fullWidth = fullWidth;
        this.captionBackground = captionBackground;
        this.captionAlpha = captionAlpha;
        this.padding = padding;
    }

    @Override
    public void apply(ImmutableImage image) {

        Graphics2D g2 = (Graphics2D) image.awt().getGraphics();
        g2.setFont(font);

        Rectangle2D bounds = g2.getFontMetrics().getStringBounds(text, g2);
        int descent = g2.getFontMetrics().getDescent();

        int captionWidth;
        if (fullWidth) {
            captionWidth = image.width;
        } else {
            captionWidth = (int) (bounds.getWidth() + padding.left + padding.right);
        }

        int captionHeight = (int) (bounds.getHeight() + padding.top + padding.bottom);

        // captionx/y are the top/left coordinates for the caption box
        int captionX, captionY;
        if (position != null) {
            Dimension dim = position.calculateXY(image.width, image.height, captionWidth, captionHeight);
            captionX = dim.getX();
            captionY = dim.getY();
        } else {
            captionX = x;
            captionY = y;
        }

        FilledRect bg = new FilledRect(
                captionX,
                image.height - captionHeight,
                captionWidth,
                captionHeight,
                g -> {
                    g.setComposite(AlphaComposite.getInstance(AlphaComposite.SRC_OVER, (float) captionAlpha));
                    g.setColor(captionBackground);
                    Graphics2DUtils.setAntiAlias(g, antiAlias);
                }
        );

        Text string = new Text(
                text,
                captionX + padding.left,
                captionY + padding.top + g2.getFontMetrics().getHeight() - descent,
                g -> {
                    g.setComposite(AlphaComposite.getInstance(AlphaComposite.SRC_OVER, (float) textAlpha));
                    g.setColor(textColor);
                    g.setFont(font);
                    Graphics2DUtils.setAntiAlias(g, antiAlias);
                });

        Canvas canvas = new Canvas(image);
        canvas.drawInPlace(bg);
        canvas.drawInPlace(string);
    }
}

